src/components/implementation/cbuf_mgr/default/cbuf_mgr.c

```diff
@@ -763,54 +847,229 @@ cbuf_unmap_at(spdid_t s_spd, unsigned int cbid, spdid_t d_spd, vaddr_t d_addr)
+	if (!f_cbi) BUG(); 				/* we literally just made this in cbuf_create so if it's not there, something is very wrong */
+	assert(f_cbi->mem);
+	memcpy_ret = memcpy(f_cbi->mem, cbi->mem, sz);
+	assert(memcpy_ret == f_cbi->mem);
+
+	/* Do a sanity check and REMOVE THIS once we're kinda confident stuff works */
+	if (memcmp(f_cbi->mem, cbi->mem, sz)) { printc("cbufs do not actually match. Womp womp :(\n"); BUG(); }
+	
+	ret = 0;
+
+	if (copy_cinfo) {
+		printc("About to try mapping cinfo cbuf into Q\n");
+		vaddr_t q_daddr = (vaddr_t)valloc_alloc(cos_spd_id(), 11, sz/PAGE_SIZE);
+		if (unlikely(!dest)) return -1;
+		flags = MAPPING_RW;
+		flags |= MAPPING_NO_VALLOC;
+		if (q_daddr != __cbuf_map_at(f_spd, f_cbid, 11, q_daddr | flags)) return -1;
+		ret = q_daddr;
+	}
+
+done:
+	return ret;
+
 }
 
-int cbuf_fork_spd(spdid_t spd, spdid_t s_spd, spdid_t d_spd)
+/*
+ * This is meant to be silly code to help me (Teo) debug LockdownOS but maybe other people would like it too.
+ * Prints out info relevant to component cci. Does NOT modify stuff.
+ */
+static void
+__get_nfo(struct cbuf_comp_info *cci)
 {
-	struct cbuf_comp_info *s_cci, *d_cci;
-	int ret = 0;
+	printc("Now iterating over global cbuf list\n");
+	// iterate over cbi_head until done
+	struct cbuf_info *current = cbi_head;
+	if (!current) {
+		printc("There should be at least one cbuf in the system\n");
+		BUG();
+	}
+
+	do {
+		// The cbuf_info list we made is actually crap because it copies the pointers so if the structure is updated it is no longer valid (how to fix???)
+		// instead, re-get the cbi from the cbuf_id which we assume is constant
+		struct cbuf_info *cbi = cmap_lookup(&cbufs, current->cbid);
 
+		if (cbi) { /* presumably the cbuf could have been deleted but this is unlikely */
+			printc("cbuf #%2u with size %x, mem start %x, and maps ", cbi->cbid, cbi->size, cbi->mem);
+			struct cbuf_maps *m = &cbi->owner;
+			struct cbuf_meta *cm = m->m;		/* Use this after loop */
+
+			do {
+				printc("[%d:%x] ", m->spdid, m->addr);
+				m = FIRST_LIST(m, next, prev);
+			} while (m != &cbi->owner);
+
+			printc("\n\tnfo: [ex %d|ow %d|t %d|in %d|re %d] page pointer %x",
+									CBUF_EXACTSZ(cm)      >> 11,
+									CBUF_OWNER(cm)        >> 10,
+									CBUF_TMEM(cm)         >>  9,
+									CBUF_INCONSISTENT(cm) >>  8,
+									CBUF_RELINQ(cm)       >>  7,
+									CBUF_PTR(cm));
+			
+			printc("\n");
+		}
+
+		current = FIRST_LIST(current, next, prev);
+	} while (current != cbi_head);
+
+	printc("Done iterating over global cbuf list\n");
+}
+
+/* This is internal so the real core of the function doesn't take a lock */
+static vaddr_t
+__cbuf_fork_spd(spdid_t o_spd, spdid_t f_spd, int cinfo_cbid)
+{
+	struct cbuf_comp_info *src, *dst;
+	struct cbuf_info *current;
+	int i;
+	vaddr_t ret = (vaddr_t) NULL;		/* address of cinfo page */
+
+	printc("starting valloc_fork_spd from %d to %d\n", o_spd, f_spd);
+	int r = valloc_fork_spd(cos_spd_id(), o_spd, f_spd);
```

Comment:
I think we all agree to remove vas_mgr.
But for valloc component, we still need discussion to choose `*_fork` or `*_at`